pkg/sshd/sshd.go

```
@@ -178,35 +124,55 @@ func Configure(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrup
-	key := p.Get("key", nil).(ssh.PublicKey)
-	return Fingerprint(key), nil
+// GenSSHKeys generates the default set of SSH host keys.
+func GenSSHKeys(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	log.Debugf(c, "Generating ssh keys for sshd")
+	// Generate a new key
+	out, err := exec.Command("ssh-keygen", "-A").CombinedOutput()
+	if err != nil {
+		log.Infof(c, "ssh-keygen: %s", out)
+		return nil, err
+	}
+	return nil, nil
 }
 
-// Fingerprint generates a colon-separated fingerprint string from a public key.
-func Fingerprint(key ssh.PublicKey) string {
-	hash := md5.Sum(key.Marshal())
-	buf := make([]byte, hex.EncodedLen(len(hash)))
-	hex.Encode(buf, hash[:])
-	// We need this in colon notation:
-	fp := make([]byte, len(buf)+15)
-
-	i, j := 0, 0
-	for ; i < len(buf); i++ {
-		if i > 0 && i%2 == 0 {
-			fp[j] = ':'
-			j++
-		}
-		fp[j] = buf[i]
-		j++
+func userInfoFromKey(key string) (*UserInfo, error) {
+	keyB64 := base64.RawURLEncoding.EncodeToString([]byte(key))
+	url := controllerURLStr("v2", "hooks", "key", keyB64)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
 	}
 
-	return string(fp)
+	builderKey, err := conf.GetBuilderKey()
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("User-Agent", "deis-builder")
+	req.Header.Add("X-Deis-Builder-Auth", builderKey)
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != 200 {
+		return nil, fmt.Errorf("user has no permissions to push in application")
+	}
+
+	ret := &UserInfo{}
+	if err := json.NewDecoder(res.Body).Decode(ret); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+func controllerURLStr(additionalPath ...string) string {
+	host := os.Getenv("DEIS_WORKFLOW_SERVICE_HOST")
+	port := os.Getenv("DEIS_WORKFLOW_SERVICE_PORT")
```

if either of these are empty, we should return an error